### PR TITLE
fix(hooks): wait for the message that closes a turn before notifying

### LINKED ISSUE #111: [Bug]: Stop notifications silently skipped for text-only replies
## Plugin version

1.40.0 (commit `dd56132`)

## Operating system

Linux (x64)

## OS version

NixOS 26.05 (Yarara)

## Terminal

Other (specify in description)

## Terminal multiplexer (if any)

_none_

## Notification type affected

Task Complete

## Issue category

Desktop notification not appearing

## What happened?

Expected: with `notifyOnTextResponse` at its default `true`, every completed turn (including quick text-only answers) should produce a "Task Complete" notification.

Actual: turns that end with a single text-only assistant reply (no tool use) are silently skipped. The debug log shows the Stop hook runs but classifies them as `unknown`:

```
[02:10:15] Analyzed status: task_complete   ← turn with tool use → notified
[02:12:42] Analyzed status: unknown         ← text-only reply → skipped
[02:13:06] Analyzed status: unknown         ← text-only reply → skipped
[02:13:42] Analyzed status: unknown         ← text-only reply → skipped
```

So the pattern from the user's perspective is: "long working turns notify, quick answers don't."

Root cause (traced through the 1.40.0 source):

1. The Stop hook often fires while Claude Code is still flushing the final assistant line to the transcript JSONL. Assistant lines can be large (thinking signatures, etc.), and for a text-only turn the *only* assistant line is written at essentially the same instant the hook runs — so the hook reads a file whose last line is truncated JSON.
2. `pkg/jsonl/jsonl.go` `Parse()` silently skips lines that fail `json.Unmarshal` (~L116-122), dropping that truncated line.
3. `FilterMessagesAfterTimestamp()` then finds zero assistant messages after the last user message, and `internal/analyzer/analyzer.go` hits the early return at L79-81:

   ```go
   if len(filteredMessages) == 0 {
       return StatusUnknown, messages, nil
   }
   ```

   `ShouldNotifyOnTextResponse()` is never consulted because that check lives below this early return.

Turns with tool use are unaffected because earlier assistant lines are already fully written by the time Stop fires, and any tool use is enough for `task_complete`.

Suggested fix: in `handleStopEvent` (or the analyzer), when the result would be `StatusUnknown` because no assistant messages follow the last user message, retry the parse once after a short delay (~150-300 ms). Optionally treat a trailing line without a terminating `\n` as "file still being written" and use that as the retry signal. A single bounded retry keeps the hook fast in the normal case and closes the race in the text-only case.

## Steps to reproduce

1. Ask Claude Code something that gets a quick one-message text answer (no tool use) → no notification, and the debug log shows `Analyzed status: unknown`.
2. Ask something that makes it run tools → notification arrives normally.

## Configuration

```json
{
  "notifications": {
    "desktop": {
      "enabled": true,
      "sound": true,
      "volume": 0.5,
      "appIcon": "${CLAUDE_PLUGIN_ROOT}/claude_icon.png",
      "clickToFocus": true,
      "terminalBundleId": ""
    },
    "webhook": { "enabled": false },
    "suppressQuestionAfterTaskCompleteSeconds": 12,
    "suppressQuestionAfterAnyNotificationSeconds": 7,
    "suppressForSubagents": true,
    "suppressFilters": []
  }
}
```

(No `notifyOnTextResponse` key — default `true` applies.)

## Error logs / output

```
[02:12:42] [DEBUG] PID:117894: === Hook triggered: Stop ===
[02:12:42] [DEBUG] PID:117894: Hook data: session=10822781-..., transcript=/home/user/.claude/projects/.../10822781-....jsonl, tool=
[02:12:42] [DEBUG] PID:117894: Analyzed status: unknown
[02:12:42] [DEBUG] PID:117894: Status is unknown, skipping notification
```

## Additional context

Desktop delivery itself works fine — the same binary, config, and daemon notify correctly when fed a fully-flushed transcript, and tool-using turns notify reliably. The issue is purely the analyzer's interaction with a mid-write transcript file.

### PR DESCRIPTION
**Pull request by liuclare**
Fixes #111.

## Problem

Claude Code appends the message that ends a turn **after** it spawns the Stop hook. The hook reads a transcript that does not yet contain the turn it was fired for.

Measured with a wall-clock probe watching both the plugin log and the transcript file — deliberately not the `timestamp` field inside the records, which is a logical time, not a write time:

| | closing message lands, relative to hook start |
|---|---|
| 13 runs | **+65 to +85 ms** |

Response length makes no difference: a three-word reply and a 600-word one are indistinguishable.

That one fact produces three separate user-visible defects.

### 1. Text-only turns never notify at all

The analyzer's window is *assistant messages after the last user message*. For a turn with no tools the closing message is the only thing that would be in it, so the window is empty, the analyzer returns `StatusUnknown` and `handleStopEvent` exits silently — every time, not intermittently. Reproduced on demand with `claude -p`: **6/6 text-only turns silent, 2/2 tool turns fine.**

This is what #111 reports: *"long working turns notify, quick answers don't."*

### 2. Tool turns notify with a generic body

A turn that called a tool classifies correctly from its `tool_use` records, which were written seconds earlier — so nothing looks wrong. But `generateTaskBody` reads the *last text* in that window, and the closing text is exactly what is missing. With no prose before the tool call, the window holds none at all and the body falls back to `Task completed successfully`.

### 3. Tool turns notify with the wrong body

The quieter half of the same defect. When prose *was* emitted before the tool call, the body is a whole, readable sentence — just the wrong one, taken from before the work happened rather than after:

```
prompt   read the first 3 lines of go.mod, and say "starting now" before you read
body     starting now                          ← what the notification showed
         The first 3 lines of go.mod are: …    ← what it should have shown
```

Nothing about that notification looks broken, which is why it survives casual use.

## Prior art

**Issue #11** (closed) reports defect 1 and was answered in v1.10.0 with the `notifyOnTextResponse` option, default on — *"notifications weren't being sent for text-only responses (when Claude responds without using any tools)."*

The option is real and it is on. On this path it is simply unreachable: `AnalyzeTranscriptWithMessages` returns `StatusUnknown` at the empty-window guard (`len(filteredMessages) == 0`), which sits many branches *ahead* of the `ShouldNotifyOnTextResponse()` call that would have promoted the turn to `task_complete`; `handleStopEvent` then skips on `StatusUnknown`. The switch can only fire for a text-only turn whose closing message is already on disk — the case that was never broken. That is how #111 came to be filed for a symptom #11 had been told was fixed.

**Issue #119** (open) is adjacent rather than duplicate, and the two interact. It asks for the opposite guard: return `StatusUnknown` when the transcript is *stale*, so a finished session stops re-notifying on a headless resume. Both are the same underlying question — whether the transcript state the hook sees is final. This PR waits for a window that is about to fill; #119 wants to reject one that will never fill again. The recency check proposed there would sit cleanly on top of this: a stale window already ends with closing text, so the settle loop exits without a single sleep and the check sees exactly the state it sees today.

## Fix

Re-read the transcript until the last record in the window carries a text block, bounded at 2s in 25ms steps — about 24× the slowest measured write, and far inside the 30s hook timeout in `hooks.json`.

Thinking blocks deliberately do not count. They are a distinct content type that never reaches the notification body, so a turn that has only thought is still unfinished. (Confirmed against real transcripts: a `Thought for 2s` before a tool call does not change the outcome.)

**The wait is skipped for statuses a tool settles on its own.** A plan or a question ends the turn on the call itself, with Claude waiting on the user rather than still writing; session limits and API errors come from the priority checks ahead of the state machine. Waiting on those would spend the whole budget for nothing.

**Waiting costs a stat per interval, not a parse.** Parsing is the expensive part — 5.8ms for a 442KB transcript, 67.8ms for a 3963KB one — so the transcript is only re-parsed once it has grown, and the ceiling is enforced by elapsed time rather than a sleep count.

## What this changes, measured over 1944 real turns

| | share | |
|---|---|---|
| tool turns that get a correct body for the first time | **48.9%** | 950 turns |
| turns that end without ever producing closing text and wait out the ceiling | **3.1%** | 60 turns |
| turns whose status changes from `task_complete` to `review_complete` | **2.5%** | 48 turns |

That last row is worth calling out explicitly: read-only turns whose long answer finally arrives now satisfy the 200-character test in rule 1d of the state machine. That is what `review_complete` was written for, but the closing text never used to be on disk in time for it to fire. Users will see 🔍 where they saw ✅ on those turns.

## Verification

End-to-end against a real Claude Code install on Linux:

- **7/7** text-only turns that were silent before now notify, settling after 75–100ms — consistent with the 65–85ms write lag plus one polling interval
- the `starting now` reproduction above now shows the closing answer instead

Seven tests, four of which fail without the fix:

| test | pins |
|---|---|
| `WaitsForTranscriptToSettle` | the turn lands mid-wait → notifies, and stops immediately (asserts exactly 2 sleeps) |
| `WaitsForClosingTextAfterTool` | tool turn with no prose → body is the answer, not `Task completed successfully` |
| `WaitsPastPreToolText` | tool turn with prose → body is the answer, not what was said before the tool |
| `SettleWaitIsBounded` | the turn never lands → no spin, no notification (asserts exactly 80 sleeps) |
| `SettleWaitHonoursElapsedTime` | transcript grows every poll → stops on elapsed time, ~21 rounds not 80 |
| `StatusEndingTheTurnDoesNotWait` | a question → **zero** sleeps |
| `SettledUnknownDoesNotWait` | `notifyOnTextResponse` off → **zero** sleeps |

They drive the wait through an injected clock where each sleep advances virtual time, so they are deterministic rather than timing-dependent.

- `go test ./...` — all packages pass
- `go test -race ./internal/hooks/` — pass
- `golangci-lint run` (v2, repo config) — 0 issues

## Notes for review

- **On the mechanism in #111.** The report traces the same defect to the same early return, but proposes that the hook reads a transcript whose last line is *truncated* JSON, which `Parse` then drops silently at [`jsonl.go#L118-L121`](https://github.com/777genius/claude-notifications-go/blob/main/pkg/jsonl/jsonl.go#L118-L121). That is mechanically sound, and I could not reproduce it: sampling as fast as I could around the hook — 309 snapshots across three turns, plus a slower probe over four more — the closing line always appeared whole. I am not claiming the report is wrong; my sampling interval is roughly 3ms and the largest single line I captured was 4316 bytes, comfortably inside the size where a write is atomic anyway. It does not change the fix either way: `Parse` drops a truncated line, so the window ends up in the same state as if the line had not been written.
- **Timeout budget.** `maxNotifyDelaySeconds` is 25 so the hook stays inside the 30s timeout. The settle wait can add up to 2s ahead of it, leaving a 3s margin instead of 5s. Typical cost is ~80ms, and a turn that never lands returns before the notify delay applies, so the two do not compound in practice.
- **Dedup is unaffected.** Phase 1 runs before analysis and phase 2 (`AcquireLock`, the `O_EXCL` acquisition that actually guards) runs after it, so the wait shifts phase 2 as a whole; concurrent duplicate hooks still leave exactly one winner.
- **Separate sleep seam.** `settleSleepFunc` rather than the existing `sleepFunc`, so tests stubbing the notify delay do not silently stub this wait too.
- The 2s / 25ms constants are unexported. Say the word if you would prefer them configurable, or a tighter ceiling — the measured worst case is 85ms, so there is a lot of headroom in 2s.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved Stop handling by waiting briefly for the assistant’s closing text to appear in the transcript before generating notification status.
  * Added bounded “transcript settling” polling that re-analyzes only while the transcript is still changing, including correct notification text for tool turns and cases where prose and `tool_use` share a message.
  * Ensured a no-wait fast path when verdict settles/unknown and text notifications are disabled.

* **Tests**
  * Expanded regression coverage for settling wait behavior, bounded termination, elapsed-time reparse costs, and multiple tool/text transcript edge cases.

### CODE REVIEW

File: internal/hooks/hooks.go [hunk 1]
```diff
@@ -534,6 +555,68 @@ func skipUTF8BOM(input io.Reader) io.Reader {
...
+		if ok && sized && size == lastSize {
+			continue
+		}
+		lastSize, sized = size, ok
+
+		status, messages, err = analyzer.AnalyzeTranscriptWithMessages(transcriptPath, h.cfg)
+		if err != nil {
+			return analyzer.StatusUnknown, nil, err
+		}
+		logging.Debug("Transcript grew, re-analyzed after %v", time.Duration(retry)*transcriptSettleInterval)
```
[the remark this answers is not in the document]
Comment (liuclare) [reply internal/hooks/hooks.go:605]: Confirmed and fixed in 5a77d6d — you are right, and this is the half my own review missed.

I had already caught the reparse cost and added a stat gate so an *unchanged* transcript is not reparsed. What I did not account for is a transcript that keeps growing without gaining an assistant record — system entries, tool results and snapshots all enlarge the file — which defeats that gate entirely and pays a full reparse on every poll.

Quantified with a virtual clock, using the 70ms reparse I measured on a 3963KB / 1019-record transcript:

| bound | rounds | simulated elapsed |
|---|---|---|
| sleep count only | 80 | ~7.6s |
| elapsed time | ~21 | 2s |

So the constant was overshooting by nearly 4x in that scenario, on top of the Stop hook's 30s budget.

The deadline now comes from a clock, with the retry count kept as a secondary bound. Tests drive `nowFunc` and `settleSleepFunc` together through a `stubSettleClock` helper: each sleep advances the virtual clock by its own argument plus a per-retry cost standing in for the reparse, so the ceiling arithmetic runs without waiting on anything.

`TestHandleStopEvent_SettleWaitHonoursElapsedTime` is the case you asked for — it grows the file on every poll and asserts the loop stops on elapsed time well before the retry count. It fails against the previous commit with `slept 80 of 80`.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

File: internal/hooks/hooks.go [hunk 2]
```diff
@@ -534,6 +556,114 @@ func skipUTF8BOM(input io.Reader) io.Reader {
...
+// are a distinct content type and deliberately do not count: they never reach the
+// notification body, so a turn that has only thought is still unfinished.
+func hasTextBlock(msg jsonl.Message) bool {
+	for _, content := range msg.Message.Content {
+		if content.Type == "text" && strings.TrimSpace(content.Text) != "" {
+			return true
+		}
+	}
+	return false
+}
```
[the remark this answers is not in the document]
Comment (liuclare) [reply internal/hooks/hooks.go:662]: Confirmed and fixed in 88b8294. I checked whether the shape occurs before changing anything — it does, but barely: **one assistant message in 15366** across the transcripts on hand carries prose and a tool call together. Rare enough to miss, cheap enough to get right.

Renamed the predicate to `messageClosesTurn`, since it no longer answers "is there prose" but "is this the answer", and it returns false on any `tool_use` regardless of what else the message holds.

Worth reporting what the fix turned up. The shared `buildTranscriptWithTools` fixture built exactly that shape, packing every tool call and the answer into a single message. 39 tests therefore described a turn that had already answered while a tool call was still pending — a state the hook cannot observe. Once the predicate was correct, each of those fixtures was unfinished at read time and sat out the full two-second ceiling on a real clock:

```
internal/hooks under -race
  before   4.6s
  after    52.1s, and TestHandleHook_DelayDoesNotHoldContentLock failing on its own 2s timeout
```

So the finding was load-bearing in a way I had not expected: not only a rare-path bug, but a fixture that had been describing an impossible turn all along. The fixture now emits one message per tool call and one for the answer, which is what Claude Code writes, and `internal/hooks` is back to 6.7s with all 21 packages green under `make test-race`.

`TestHandleStopEvent_WaitsWhenTextSharesMessageWithToolUse` pins the shape; it fails against the previous commit with the preamble as the body.

🤖 Generated with [Claude Code](https://claude.com/claude-code)
